File: Commons/Graph/src/test/java/com/greentree/common/graph/LcaTest.java
package com.greentree.common.graph;

import com.greentree.commons.graph.RootTree;
import com.greentree.commons.graph.RootTreeBase;
import org.junit.jupiter.api.Test;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class LcaTest {

    @Test
    void test1() {
        final var tree = new RootTreeBase<>("A");
        tree.add("B", "A");
        tree.add("C", "A");
        tree.add("D", "B");
        tree.add("E", "B");
        final var a = tree.lca("D", "C");
        assertEquals(a, "A");
    }

    @Test
    void classTree() {
        final var tree = new RootTree<Class<?>>() {

            @Override
            public Iterator<Class<?>> iterator() {
                return null;
            }

            @Override
            public boolean containsInSubTree(Object v, Object parent) {
                var cls = (Class<?>) v;
                var superClass = (Class<?>) parent;
                return superClass.isAssignableFrom(cls);
            }

            @Override
            public Class<?> getParent(Object v) {
                if (v == Object.class)
                    return Object.class;
                return ((Class<?>) v).getSuperclass();
            }
        };
        final var a = tree.lca(ArrayList.class, LinkedList.class);
        assertEquals(a, AbstractList.class);
    }

}

File: Commons/Graph/src/test/java/com/greentree/common/graph/LcaTest.kt
package com.greentree.common.graph

import com.greentree.commons.graph.RootTree
import com.greentree.commons.graph.RootTreeBase
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.util.*

@Disabled
class LcaTest {

	@Test
	fun test1() {
		val tree = RootTreeBase("A")
		tree.add("B", "A")
		tree.add("C", "A")
		tree.add("D", "B")
		tree.add("E", "B")
		val a = tree.lca("D", "C")
		Assertions.assertEquals(a, "A")
	}

	@Test
	fun classTree() {
		val tree: RootTree<Class<*>> = object : RootTree<Class<*>> {
			override fun containsInSubTree(cls: Class<*>, superClass: Class<*>): Boolean {
				return superClass.isAssignableFrom(cls)
			}

			override fun getParent(v: Class<*>): Class<*> {
				if(v === Any::class.java) return Any::class.java
				return (v as Class<*>?)!!.superclass
			}

			override fun iterator(): Iterator<Class<*>> {
				TODO("Not yet implemented")
			}
		}
		val a = tree.lca(ArrayList::class.java, LinkedList::class.java)
		Assertions.assertEquals(a, AbstractList::class.java)
	}
}
